Rosa Bulo (REB) SCMSUITE-- SO107: MDAnalysisJobs now Pisa compatible

rebadam opened the pull request:
- The MDAnalysisJobs should now work with Pisa instead of Settings
- It is possible to add multiple previous_jobs (in the case of time analysis, it is up to the user to ensure they are consecutive)

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: MDAnalysisJobs now Pisa compatible

- The MDAnalysisJobs should now work with Pisa instead of Settings
- It is possible to add multiple previous_jobs
  (in the case of time analysis, it is up to the user to ensure they are consecutive)

mhellstr commented on recipes/md/trajectoryanalysis.py:
```diff
@@ -30,19 +36,109 @@ def _get_max_dt_frames(self, max_correlation_time_fs):
+        if self.atom_indices and self._parent_write_atoms:
+            section = getattr(self.settings.input, self._task)
+            for entry in section:
+                found = self._has_settings_entry(entry, "Atoms")
+                if found:
+                    found = self._has_settings_entry(entry.Atoms, "Atom")
```
Better to combine these with an "and"

mhellstr commented:
In general this code is quite difficult to read, isn't it easier to just convert the PISA object to Settings once and be done with it? Similar to how RD converts from settings-to-pisa, but the other way around.

dormrod commented on unit_tests/test_tools_plot.py:
```diff
@@ -393,6 +394,54 @@ def test_plot_msd(run_calculations, rkf_tools_plot, xyz_folder):
+    sets.MeanSquareDisplacement.Property = "Coords"
+    sets.MeanSquareDisplacement.Atoms.Element = "O"
+    s_msd = Settings()
+    s_msd.input = sets
```
In the `AMSJob` the input class from PISA can either fully replace the settings object, or be under `settings.input`. Does the former also work here?

Answered my own question - yes as this happens in the `Job` base class, but maybe also cover in the test somehow? i.e. using  `AMSMSDJob(job, settings=sets)`

rebadam replied:
Ah, I did not know that. Will add it.

dormrod commented on interfaces/adfsuite/amsanalysis.py:
```diff
@@ -288,14 +298,19 @@ def _serialize_mol(self):
         systems = AMSJob._serialize_molecule(self)
         if len(systems) > 0:
-            self.settings.input.system = systems
+            if _has_scm_pisa and isinstance(self.settings.input, DriverBlock):
```
More for my knowledge, but do we know why this uses `SCMJob` instead of `AMSJob`?

Would we get this all "for free" if we refactored to use `AMSJob` and/or would it cause any issues?

rebadam replied:
I am not sure if it would make sense to use AMSJob as the parent. I think the main difference is that it does not produce a .rkf file, but a .kf file. And it does not need an engine block. Intuitively I don't think inheritance from AMSJob makes sense, but it is possible that it makes practical sense.

dormrod replied:
OK thanks! I assumed `AMSXJob` would inherit from `AMSJob` and `SCMJob` was a legacy version of the `AMSJob`, but seems like it is still useful

dormrod commented on recipes/md/trajectoryanalysis.py:
```diff
@@ -30,19 +36,109 @@ def _get_max_dt_frames(self, max_correlation_time_fs):
+            prevjobs = [prevjobs]
+
+        # Read history and time step
+        # FIXME: Does not account for the presence of Range of StepSize (which the user may have provided)
```
Fix needed here?

rebadam replied:
Yes, but this is a separate issue. It should be fixed, but maybe not in this branch? It is already doing much more than I planned.

dormrod replied:
Fair enough!

dormrod commented on recipes/md/trajectoryanalysis.py:
```diff
@@ -30,19 +36,109 @@ def _get_max_dt_frames(self, max_correlation_time_fs):
+        """
+        Convert the main settings object for this Task to a list of settings objects
+        """
+        if isinstance(self.settings.input, Settings):
```
To Matti's point, for this class it might just be easier to use:

```
        if _has_scm_pisa and isinstance(self.settings.input, DriverBlock):
            self.settings.input = self.settings.input.to_settings()
```

or similar in the c'tor, to remove all the if/else.

Trade off would be that the user would not get the benefits of PISA when accessing the job settings...

rebadam replied:
Yes, that is definitely an option. But it would be different from the AMSJob approach. Wouldn't it be better to make the pisa interface more compatible with the Settings object? For example, give it a get() method, and maybe __contains__? There is also a bug in Pisa for non-unique entries. If you provide a single entry first, then you cannot easily add other entries. Right now I just added work-around code, but I think it should be fixed in Pisa.

dormrod replied:
Yes I think PISA should be improved. Could you please raise JIRA(s) with the examples of the bugs/methods to add?

In particular I agree the `contains` is a big omission at the moment and should behave as the `Settings` contains.

I also think that the PISA/Settings objects should be more generally interchangeable to facilitate this kind of refactor. In an ideal world you shouldn't have to make this change at all... (or at least it should be very minimal)

dormrod commented:
Also please update: https://github.com/SCM-NV/PLAMS/blob/trunk/CHANGELOG.md

under `Unreleased/Added`

dormrod committed:
Fix job loading for dill files from PLAMS<2025

dormrod committed:
Update changelog

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107:Updated AnalysisJobs Pisa compatibility

- Made changes suggested by David and Matti
- Compacted the PISA code more - Localized the if-statements in only a few methods.

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: Updated changelog

rebadam commented:
I don't really understand why the above is even a conflict. David, could you maybe take a look?

dormrod committed:
Merge branch 'trunk' into mdjobs
